examples/custom-assessments/src/main.py

```diff
@@ -0,0 +1,211 @@
+        # IMPORTANT: Change to import your own environment
+        self.sim = SimEnv(env_config={})
+
+    def episode_to_df(
+        self, states, actions, rewards, terminated, truncated, n_steps
+    ) -> pd.DataFrame:
+        """Create Pandas dataframe from episode steps.
+
+        Args:
+            states (dict): The states of the simulation at each step.
+            actions (dict): The actions of the agent at each step.
+            rewards (list): The rewards of the agent at each step.
+            terminated (list): Whether the episode terminated at each step.
+            truncated (list): Whether the episode was truncated at each step.
+            n_steps (int): The final number of steps in the episode.
+
+        Returns:
+            pd.DataFrame: The dataframe containing the states, actions, episode id,
+            step number, and initial condition.
+        """
+        df = pd.concat([pd.DataFrame(dict) for dict in [states, actions]], axis=1)
+        # Prepend "state_" to state columns and "action_" to action columns for clarity
+        df.rename(
+            columns=lambda x: "action_" + x if x in actions else "state_" + x,
+            inplace=True,
+        )
+        for key, value in self.initial_condition.items():
+            df["config_" + key] = value
+        df["reward"] = pd.Series(rewards)
+        df["reward_cumsum"] = df["reward"].cumsum()
+        df["terminated"] = pd.Series(terminated)
+        df["truncated"] = pd.Series(truncated)
+        df["episode_id"] = self.episode_id
+        df["step_id"] = pd.Series(np.arange(n_steps, dtype=int))
+
+        return df
+
+    def run_episode(
+        self,
+        initial_condition: Dict[str, int],
+        episode_id: int = 0,
+        max_steps: int = 100,
+    ) -> pd.DataFrame:
+        """Run an episode for the initial conditions using the agent policy and the
+        simulator.
+
+        Args:
+            initial_condition (dict): The initial condition for the simulation.
+            episode_id (int, optional): The id of the episode. Defaults to 0.
+            max_steps (int, optional): The maximum number of steps for the
+                episode. Defaults to 100.
+
+        Returns:
+            pd.DataFrame: The dataframe containing the results of the episode run.
+        """
+        # assign values to object properties
+        self.initial_condition = initial_condition
+        self.episode_id = episode_id
+        self.max_steps = max_steps
+
+        # Reset state with initial condition and return observable state
+        observable_state, _ = self.sim.reset(options=self.initial_condition)
+
+        # Initialize lists for episode logging
+        states = []
+        actions = []
+        rewards = []
+        terminated = []
+        truncated = []
+        # Test for max_steps
+        for step in range(self.max_steps):
+            # Get full current state and save it
+            full_state = self.sim.state.copy()
```

Comment:
Could you say why you're `copy`-ing the state? It does not seem you're making modifications below.
Also, if users will put a list as state, or a simple number would this work too or fail?

Reply:
This was updated slightly in the latest commit, but it's there because self.sim.state can be a dictionary (which is mutable). Stepping the sim will change all the state of all previous steps to the current state.

Reply:
The code in yesterday's commit addresses the list issue.